crypto/rand/rand_lib.c

```diff
@@ -49,13 +72,19 @@ const RAND_METHOD *RAND_get_rand_method(void)
 #endif
             default_RAND_meth = RAND_OpenSSL();
     }
-    return default_RAND_meth;
+    tmp_meth = default_RAND_meth;
```

Comment:
I believe that the function call to unlock will provide a compiler-level barrier (and the lock itself will give a memory barrier at runtime).  Function calls are considered to have side effects, in the abstract machine.